### ISSUE TITLE
Proposal: option to disable implicit boolean cast in conditions

### ISSUE THREAD

**Issue by impworks** [Microsoft/TypeScript#19191]
I came across a very silly bug in my code which the Typescript compiler did not catch. Consider the following snippet of code:

```
class MyClass {
    public /*get*/ flag(): boolean {
        return false;
    }

    public doStuff() {
        if(this.flag) {
            alert('hello');
        }
    }
}
```

The `alert` will always be executed, because the property has been mis-declared as a method. Instead of invoking the getter and checking its return value, we instead evaluate the *body* as truthy.

While this behaviour is expected in many cases, it would be nice to have a compilation flag to enable additional checks, for example - `noImplicitBool`. It would require expressions in classic control structures (`if`, `while`, `for`) to be explicitly of the `boolean` type.

The flag should not affect the behaviour of the `||` operator, which is often used for coalescing values of arbitrary types.

**Comment by aluanhaddad** [comment]
See #7746

**Comment by ajafff** [comment]
TSLint has a rule `strict-boolean-expressions` that does exactly what you want: https://palantir.github.io/tslint/rules/strict-boolean-expressions/

**Comment by RyanCavanaugh** [comment]
See #9041

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
